### PR TITLE
Make reliability layer parameters message specific - message specific approach.

### PR DESCRIPTION
**Pull request by boaks**
Based on Lari Hotari work in PR #1069 this implements the reliability layer parameters message specific.

Signed-off-by: Achim Kraus <achim.kraus@bosch-si.com>

### CODE REVIEW

**Comment by boaks** [comment]
@lhotari 

during the discussion of your pull request #1069 I proposed to solve that function based on passing the parameters with the message. For me it simplifies it, though it doesn't require the "endpoint specific EndpointSpecificConfigProvider" stuff. OK, it mainly push that out of the scope of the stack into the application responsibility.

We will see, if you and the other committers agree on this or your original PR.

If this PR gets agreed, it would be great, if you change your PR into a PR containing only your `ReliabilityLayerParameters`. Please ensure, that the file has a license header. if you contribute as private person, then you may just copy the header of `ReliabilityLayerParameters` here in this PR. Otherwise, please replace the name with the name of your company into that license header. Thanks in advance!

File: californium-core/src/main/java/org/eclipse/californium/core/network/RemoteEndpoint.java [hunk 1]
```diff
@@ -197,7 +199,11 @@ public void setEstimatorValues(long rto, long rtt, long rttvar, int estimatorTyp
 	public Queue<Exchange> getNonConfirmableQueue(){
 		return nonConfirmableQueue;
 	}
-	
+
+	public ReliabilityLayerParameters getConfigForEndpoint() {
+		return endpointReliabilityLayerParameters;
+	}
```
Comment (sbernard31) [inline californium-core/src/main/java/org/eclipse/californium/core/network/RemoteEndpoint.java]: "ForEndpoint" is not needed ? and I find getConfig to general ?

Why not something like getReliabilityLayerParameters() ?

by the way the specficiation seems to call that : "Transmission Parameters"
(see https://tools.ietf.org/html/rfc7252#section-4.8 )

File: californium-core/src/main/java/org/eclipse/californium/core/network/stack/ReliabilityLayerParameters.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+		}
+
+		public Builder applyConfig(NetworkConfig config) {
+			ackTimeout = config.getInt(NetworkConfig.Keys.ACK_TIMEOUT);
+			ackRandomFactor = config.getFloat(NetworkConfig.Keys.ACK_RANDOM_FACTOR);
+			ackTimeoutScale = config.getFloat(NetworkConfig.Keys.ACK_TIMEOUT_SCALE);
+			maxRetransmit = config.getInt(NetworkConfig.Keys.MAX_RETRANSMIT);
+			nstart = config.getInt(NetworkConfig.Keys.NSTART);
+			return this;
+		}
```
Comment (sbernard31) [inline californium-core/src/main/java/org/eclipse/californium/core/network/stack/ReliabilityLayerParameters.java]: I still not like so much the idea to make this class dependent of NetworkConfig but this time there is not easy alternative :/

File: californium-core/src/main/java/org/eclipse/californium/core/network/stack/ReliabilityLayerParameters.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+ *     Lari Hotari - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.californium.core.network.stack;
+
+import org.eclipse.californium.core.network.config.NetworkConfig;
+
+/**
+ * Contains the configuration values for {@link ReliabilityLayer} and {@link CongestionControlLayer}
+ */
+public class ReliabilityLayerParameters {
```
Comment (sbernard31) [inline californium-core/src/main/java/org/eclipse/californium/core/network/stack/ReliabilityLayerParameters.java]: should we call this class TransmissionParameters ? and add a reference to this url in the comment https://tools.ietf.org/html/rfc7252#section-4.8 ?

**Comment by boaks** [comment]
@lhotari 

Though we want to release 2.0.0-M18 as RC1, I would appreciate, if you can provide us your feedback.

**Comment by boaks** [comment]
@sbernard31 

According emo-ip-team, we merge PR #1069 and then overwrite that by this PR.
Do you generally agree?

if so, I will go to  rename the function to getReliabilityLayerParameters()  according your comment and merge the PRs.

**Comment by sbernard31** [comment]
Do you generally agree?

:+1:

**Comment by boaks** [comment]
merged with commit 880fe965cac302a4e5b827c287b8f6324afd7b11

**Comment by lhotari** [comment]
> Though we want to release 2.0.0-M18 as RC1, I would appreciate, if you can provide us your feedback.

Thank you @boaks . This looks great. It simplifies the solution a lot when the reliability parameters can be set at message level.

**Pull request closed without merging**